File: testSuite/test-noninstantaneous-recycling.pl
#!/usr/bin/env perl
use strict;
use warnings;
use lib $ENV{'GALACTICUS_EXEC_PATH'         }."/perl";
use lib $ENV{'GALACTICUS_ANALYSIS_PERL_PATH'}."/perl";
use PDL;
use PDL::NiceSlice;
use PDL::IO::HDF5;
use File::Slurp qw(slurp);
use System::Redirect;

# Check calculations of noninstantaneous recycling.
# Andrew Benson (23-September-2023)

# Run the model.
&System::Redirect::tofile("mkdir -p outputs; cd ..; ./Galacticus.exe testSuite/parameters/noninstantaneous_recycling.xml","outputs/noninstantaneous_recycling.log");
unless ( $? == 0 ) {
    print "FAILED:  model run:\n";
    system("cat outputs/noninstantaneous_recycling.log");
} else {
    print "SUCCESS: model run\n";
}

my $yields;
open(my $logFile,"outputs/noninstantaneous_recycling.log");
while ( my $line = <$logFile> ) {
    if ( $line =~ m/(\S+\/yield(Metals|Fe)_[a-z0-9]+.hdf5)/ ) {
	my $fileName               = $1;
	my $type                   = $2;
	$yields->{'file'}->{$type} = $fileName;
	my $file                   = new PDL::IO::HDF5($fileName);
	$yields          ->{$type} = $file->dataset("yield".$type)->get();
    }
}
close($logFile);
unless ( exists($yields->{'Metals'}) && exists($yields->{'Fe'}) ) {
    foreach my $type ( "Metals", "Fe" ) {
	if ( exists($yields->{'file'}->{$type}) ) {
	    print "found ".$type." yield in '".$yields->{'file'}->{$type}."'\n";
	} else {
	    print "failed to find file for ".$type." yield\n";
	    print "log file is:\n";
	    system("cat outputs/noninstantaneous_recycling.log");
	}
    }
    die("failed to read yield tables");
}

my $haveMetals   = which($yields->{'Metals'} > 0.0);
my $ratioMaximum = maximum(
                           +$yields->{'Fe'    }->flat()->($haveMetals)
                           /$yields->{'Metals'}->flat()->($haveMetals)
                          );

# Read the model data and check for consistency.
my $model   = new PDL::IO::HDF5("outputs/noninstantaneous_recycling.hdf5");
my $outputs  = $model  ->group('Outputs' );
my $output   = $outputs->group('Output10');
my $nodeData = $output ->group('nodeData');
my $data;
$data->{$_} = $nodeData->dataset($_)->get()
    foreach ( 
	"spheroidAbundancesStellarMetals",
	"spheroidAbundancesStellarFe"    ,
	"diskAbundancesStellarMetals"    ,
	"diskAbundancesStellarFe"
    );
my $massMetals = $data->{'spheroidAbundancesStellarMetals'}+$data->{'diskAbundancesStellarMetals'};
my $massFe     = $data->{'spheroidAbundancesStellarFe'    }+$data->{'diskAbundancesStellarFe'    };
my $nonZero    = which($massMetals > 1.0);
my $ratio      = +$massFe    ->($nonZero)
                 /$massMetals->($nonZero); 
my $status = all($ratio < $ratioMaximum) ? "SUCCESS" : "FAILED";
print $status.": Fe/Z ratio\n";
exit 0;

File: testSuite/test-noninstantaneous-recycling.py
#!/usr/bin/env python3
import subprocess
import sys
import re
import h5py
import numpy as np

# Check calculations of noninstantaneous recycling.
# Andrew Benson (ported to Python)

# Run the model.
subprocess.run("mkdir -p outputs", shell=True)
with open("outputs/noninstantaneous_recycling.log", "w") as logFile:
    status = subprocess.run(
        "cd ..; ./Galacticus.exe testSuite/parameters/noninstantaneous_recycling.xml",
        shell=True, stdout=logFile, stderr=subprocess.STDOUT
    )
if status.returncode != 0:
    print("FAILED:  model run:")
    with open("outputs/noninstantaneous_recycling.log") as f:
        print(f.read())
else:
    print("SUCCESS: model run")

yields = {}
with open("outputs/noninstantaneous_recycling.log") as logFile:
    for line in logFile:
        m = re.search(r'(\S+/yield(Metals|Fe)_[a-z0-9]+\.hdf5)', line)
        if m:
            fileName = m.group(1)
            ytype    = m.group(2)
            yields.setdefault('file', {})[ytype] = fileName
            with h5py.File(fileName, "r") as f:
                yields[ytype] = f["yield" + ytype][:]

if "Metals" not in yields or "Fe" not in yields:
    for ytype in ("Metals", "Fe"):
        if ytype in yields.get("file", {}):
            print(f"found {ytype} yield in '{yields['file'][ytype]}'")
        else:
            print(f"failed to find file for {ytype} yield")
            print("log file is:")
            with open("outputs/noninstantaneous_recycling.log") as f:
                print(f.read())
    print("FAILED: could not read yield tables")
    sys.exit(0)

haveMetals   = np.where(yields["Metals"].flat[:] > 0.0)[0]
ratioMaximum = np.max(yields["Fe"].flat[haveMetals] / yields["Metals"].flat[haveMetals])

# Read the model data and check for consistency.
with h5py.File("outputs/noninstantaneous_recycling.hdf5", "r") as model:
    nodeData  = model["Outputs/Output10/nodeData"]
    dataNames = [
        "spheroidAbundancesStellarMetals",
        "spheroidAbundancesStellarFe",
        "diskAbundancesStellarMetals",
        "diskAbundancesStellarFe",
    ]
    data = {name: nodeData[name][:] for name in dataNames}

massMetals = data["spheroidAbundancesStellarMetals"] + data["diskAbundancesStellarMetals"]
massFe     = data["spheroidAbundancesStellarFe"]     + data["diskAbundancesStellarFe"]
nonZero    = np.where(massMetals > 1.0)[0]
ratio      = massFe[nonZero] / massMetals[nonZero]
status_str = "SUCCESS" if np.all(ratio < ratioMaximum) else "FAILED"
print(f"{status_str}: Fe/Z ratio")
